src/Components/Forms/src/FieldIdentifier.cs

```diff
@@ -100,42 +100,69 @@ private static void ParseAccessor<T>(Expression<Func<T>> accessor, out object mo
-        if (memberExpression.Expression is ConstantExpression constantExpression)
-        {
-            if (constantExpression.Value is null)
+            // Get a reference to the model object
+            // i.e., given a value like "(something).MemberName", determine the runtime value of "(something)",
+            if (memberExpression.Expression is ConstantExpression constantExpression)
+            {
+                if (constantExpression.Value is null)
+                {
+                    throw new ArgumentException("The provided expression must evaluate to a non-null value.");
+                }
+                model = constantExpression.Value;
+            }
+            else if (memberExpression.Expression != null)
+            {
+                // It would be great to cache this somehow, but it's unclear there's a reasonable way to do
+                // so, given that it embeds captured values such as "this". We could consider special-casing
+                // for "() => something.Member" and building a cache keyed by "something.GetType()" with values
+                // of type Func<object, object> so we can cheaply map from "something" to "something.Member".
+                var modelLambda = Expression.Lambda(memberExpression.Expression);
+                var modelLambdaCompiled = (Func<object?>)modelLambda.Compile();
+                var result = modelLambdaCompiled();
+                if (result is null)
+                {
+                    throw new ArgumentException("The provided expression must evaluate to a non-null value.");
+                }
+                model = result;
+            }
+            else
             {
                 throw new ArgumentException("The provided expression must evaluate to a non-null value.");
             }
-            model = constantExpression.Value;
         }
-        else if (memberExpression.Expression != null)
+        else if (accessorBody is MethodCallExpression methodCallExpression
+            && methodCallExpression.Method.IsSpecialName
+            && methodCallExpression.Method.Name == "get_Item")
         {
-            // It would be great to cache this somehow, but it's unclear there's a reasonable way to do
-            // so, given that it embeds captured values such as "this". We could consider special-casing
-            // for "() => something.Member" and building a cache keyed by "something.GetType()" with values
-            // of type Func<object, object> so we can cheaply map from "something" to "something.Member".
-            var modelLambda = Expression.Lambda(memberExpression.Expression);
-            var modelLambdaCompiled = (Func<object?>)modelLambda.Compile();
-            var result = modelLambdaCompiled();
-            if (result is null)
+            // Field name will be the index value, enclosed in brackets
+            fieldName = $"[{methodCallExpression.Arguments[0]}]";
```

Comment:
I'm not 100% sure what will happen here. How do you know that `methodCallExpression.Arguments[0]` even has a meaningful string representation?

Does this only work if the indexer keys happen to be primitives like ints/strings, or would it also work if the indexer keys are arbitrary objects whose string representations are just their type name?

Reply:
To be more precise, consider scenarios like:

```razor
@foreach (var line in order.Lines)
{
    <ValidationMessage For="@(() => lineInfos[line])" />
}
```

... where `lineInfos` is a `Dictionary<OrderLine, OrderLineInfo>`.

Reply:
Also, we should bear in mind that whatever format is used for the fieldname here is not just an internal implementation detail. It becomes part of the public API contract, and has to be documented, not changed in the future, etc.. That's because people will have to write things like `new FieldIdentifier(someObject, "[15]")` if they are trying to access the validation state directly.

This wasn't an issue in the past because there's nothing ambiguous about a property name (so, `new FieldIdentifier(customer, "PhoneNumber")` is clear and obvious). But if we make up some convention around wrapping things in square brackets, that becomes a convention we have to document and support.